lib/Dialect/Moore/Transforms/VectorizationPass.cpp

```diff
@@ -0,0 +1,134 @@
+#include "mlir/IR/Builders.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/Support/raw_ostream.h"
+#include "circt/Dialect/Moore/MooreOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Pass/PassRegistry.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Support/IndentedOstream.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+#include "llvm/Config/llvm-config.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace moore;
+
+namespace {
+
+struct ScalarAssignGroup {
+  moore::ExtractRefOp extractRef;
+  moore::ExtractOp extract;
+  moore::ContinuousAssignOp assign;
+  int index;
+};
+
+struct ValueComparator {
+  bool operator()(mlir::Value lhs, mlir::Value rhs) const {
+    return lhs.getAsOpaquePointer() < rhs.getAsOpaquePointer();
+  }
+};
+
+using IndexedGroupMap = std::map<int, ScalarAssignGroup>;
+using SourceGroupMap = std::map<mlir::Value, IndexedGroupMap, ValueComparator>;
+using AssignTree = std::map<mlir::Value, SourceGroupMap, ValueComparator>;
+
+void vectorizeGroup(std::vector<ScalarAssignGroup> &group) {
+  if (group.empty())
+    return;
+
+  auto builder = OpBuilder(group.front().assign.getContext());
+  builder.setInsertionPoint(group.front().assign);
+
+  auto dst = group.front().extractRef.getOperand();
+  auto src = group.front().extract.getOperand();
+
+  builder.create<moore::ContinuousAssignOp>(group.front().assign.getLoc(), dst, src);
+
+  for (auto &g : group) {
+    g.assign.erase();
+    g.extractRef.erase();
+    g.extract.erase();
+  }
```

Comment:
Before erasing the extract ops we have to check whether there are other users of the SSA value they define. Otherwise, this would lead to a crash.